language-service/src/services/yamlDefinition.ts

```diff
@@ -0,0 +1,67 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Red Hat, Inc. All rights reserved.
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+'use strict';
+
+import { YAMLDocument } from "../parser/yamlParser";
+import { PromiseConstructor, Thenable } from "vscode-json-languageservice";
+import { TextDocument, Position, Definition, Location } from "vscode-languageserver-types";
+import { URI, Utils } from "vscode-uri";
+
+export class YAMLDefinition {
+
+    private promise: PromiseConstructor;
+
+    constructor(promiseConstructor: PromiseConstructor) {
+        this.promise = promiseConstructor || Promise;
+    }
+
+    public doDefinition(document: TextDocument, position: Position, yamlDocument: YAMLDocument, workspaceRoot: URI): Thenable<Definition> {
+        const offset = document.offsetAt(position);
+
+        const jsonDocument = yamlDocument.documents.length > 0 ? yamlDocument.documents[0] : null;
+        if(jsonDocument === null){
+            return this.promise.resolve(void 0);
+        }
+        let node = jsonDocument.getNodeFromOffset(offset);
+        if (!node || (node.type === 'object' || node.type === 'array') && offset > node.start + 1 && offset < node.end - 1) {
+            return this.promise.resolve(void 0);
+        }
+        
+        // can only jump to definition for template declaration
+        if (node.location !== 'template') {
+            return this.promise.resolve(void 0);
+        }
+
+        const [location, resource] = node
+            .getValue()
+            .split('@');
+
+        // cannot jump to external resources
+        if (resource && resource != 'self') {
+            return this.promise.resolve(void 0);
+        }
+
+        // determine if abs path (from root) or relative path
+        let pathToDefinition = '';
+        if (location.startsWith('/')) {
```

Comment:
Is `\absolute\windows\path\with\backslashes` also supported? (As weird as that looks)

Guess that's something for me to play around with.

Reply:
According to the ADO pipelines documentation:

> Template paths can be an absolute path within the repository or relative to the file that does the including.
>
> To use an absolute path, the template path must start with a `/`. All other paths are considered relative

https://learn.microsoft.com/en-us/azure/devops/pipelines/process/templates?view=azure-devops&pivots=templates-includes#reference-template-paths

I see no reference at all to using the backslash `\` for pathing, so my assumption is that it isn't valid and would probably be read as escape chars.

Reply:
Wow that's pretty interesting. I supposed that `vscode-uri` knows how to deal with the backslashes then. In which case I can just add a check to see if the path starts with `/` or `\` to know if it's absolute or relative

Reply:
So it seems as though `vscode-uri` is not very good with the backslash for paths. I did this little example to mimic how it would look in the scenario here

```[javascript]
const { URI, Utils } = require('vscode-uri')
const workspaceRoot = URI.parse('/home/repo')
const location = '\\pipelines\\templates\\template.yml'

const pathToDefinition = Utils.joinPath(workspaceRoot, location).fsPath
console.log(pathToDefinition) // '/home/repo/\\pipelines\\templates\\template.yml' -> not what we want
```

So I'm thinking we need to either ignore the weird backslash scenario, or use something like `location.replace(/\\/g, '/')` to replace the backslashes. But I feel this comes with some caveats as well

Reply:
> but in azure pipelines template references it is always a path seperator

Thank goodness. What a headache that would have been if templates behaved differently depending on what OS you were running your build on.

As for this:
> So I'm thinking we need to either ignore the weird backslash scenario

I do think your proposed "replace all" solution would work (but maybe instead of "/", use `path.sep` instead). What are the caveats you're referring to?

Reply:
(Btw, in response to "I'm working on another vscode extension, which can simulate template expansion locally", I would gladly take that as a contribution into azure-pipelines-vscode, if you're interested!)